migratekvdb/migration.go

```diff
@@ -0,0 +1,1296 @@
+	ChunkSize uint64
+
+	// Logger is the logger to use for logging.
+	Logger btclog.Logger
+
+	// ForceNewMigration is a flag to force a new migration even if a
+	// migration state is found. This will delete all existing migration
+	// data in the target database.
+	ForceNewMigration bool
+}
+
+// validateConfig ensures the configuration is valid and sets defaults.
+func validateConfig(cfg *Config) error {
+	if cfg == nil {
+		return fmt.Errorf("config cannot be nil")
+	}
+
+	// Validate and set defaults for chunk size.
+	switch {
+	case cfg.ChunkSize == 0:
+		cfg.ChunkSize = DefaultChunkSize
+
+		cfg.Logger.Infof("Chunk size set to default: %d bytes",
+			cfg.ChunkSize)
+
+	case cfg.ChunkSize > MaxChunkSize:
+		return fmt.Errorf("chunk size too large: %d bytes, maximum "+
+			"is %d bytes", cfg.ChunkSize, MaxChunkSize)
+	}
+
+	// Ensure logger is set.
+	if cfg.Logger == nil {
+		cfg.Logger = btclog.Disabled
+	}
+
+	return nil
+}
+
+// ChunkHash represents the verification hash of a chunk of the source database.
+type ChunkHash struct {
+	// Hash is the accumulated hash of the data which is included in one
+	// chunk.
+	Hash uint64 `json:"hash"`
+
+	// LastKey is the last key processed in the chunk so we can resume
+	// the migration from the same point.
+	LastKey []byte `json:"last_key"`
+
+	// Path is the last processed bucket path in this chunk.
+	Path BucketPath `json:"path"`
+}
+
+// Equal compares this ChunkHash with another ChunkHash for equality.
+func (c *ChunkHash) Equal(other *ChunkHash) bool {
+	// Check for nil
+	if c == nil || other == nil {
+		return c == other
+	}
+
+	if c.Hash != other.Hash {
+		return false
+	}
+
+	if !bytes.Equal(c.LastKey, other.LastKey) {
+		return false
+	}
+
+	return c.Path.Equal(other.Path)
+}
+
+// String returns a string representation of the ChunkHash.
+func (c *ChunkHash) String() string {
+	if c == nil {
+		return "ChunkHash<nil>"
+	}
+	return fmt.Sprintf("ChunkHash{hash=%d, lastKey=%x, path=%v}",
+		c.Hash, c.LastKey, c.Path)
+}
+
+// Migrator handles the chunked migration of bolt databases. It supports:
+//   - Resumable migrations through state tracking
+//   - Chunk-based processing to handle large databases
+//   - Verification of migrated data (also resumable)
+//   - Progress logging
+type Migrator struct {
+	// cfg is the configuration for the migrator.
+	cfg Config
+
+	// migration is the migration state which is used to track and also
+	// resume the migration.
+	migration *MigrationState
+
+	// verification is the verification state which is used to track and
+	// also resume the verification.
+	verification *VerificationState
+}
+
+// New creates a new Migrator with the given configuration.
+func New(cfg Config) (*Migrator, error) {
+	// Validate and set defaults for the config.
+	if err := validateConfig(&cfg); err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	return &Migrator{
+		cfg: cfg,
+		migration: &MigrationState{
+			hash:     fnv.New64a(),
```

Comment:
I think it'd be safer to use the 128-bit version here (lower collision probability). We're I/O bound, so it shouldn't affect our performance too much.

Reply:
removed the hash verification and now just comparing all the values individually.